Threading refactoring

msimberg opened the pull request:
Minor threading refactoring to remove a few more dependencies.

- Use `hpx::threads::detail::get_thread_num_tss` instead of `hpx::get_worker_thread_num` to get thread number in schedulers etc.
- Pass thread pool into `hpx::thread` constructor. Old constructors remain, but don't fall back to the default thread pool if not on an HPX thread.
- Remove some leftover includes.

msimberg commented on src/runtime/threads/thread.cpp:
```diff
@@ -36,17 +34,11 @@ namespace hpx
-        catch(...) {
-            hpx::report_error(std::current_exception());
-            /* nothing else we can do */;
-        }
```
Great, thanks! I'll check.

msimberg replied:
I can't see why we would need this anymore. I've added a test which throws from `thread` (`join` a thread twice) which required releasing some locks but otherwise works well.

hkaiser commented on src/runtime/threads/thread.cpp:
```diff
@@ -77,11 +61,10 @@ namespace hpx
         {
-            terminate("thread::~thread", "destroying running thread");
+            HPX_THROW_EXCEPTION(
+                invalid_status, "thread::~thread", "destroying running thread");
```
We are not allowed to throw exceptions in destructors...

msimberg replied:
Argh, good point... thanks.

msimberg commented:
I added a special handler to call when something goes wrong in the `thread` destructor but the other changes are fairly straightforward. As a flyby I moved the handlers from `hpx_init.cpp` to a separate file.

biddisco commented:
I'm happy to see it merged, the sooner all this thread related stuff goes in, the sooner I can prepare my scheduler changes for a PR

msimberg commented:
@biddisco we're not merging this yet since the change to `hpx::thread` adds a constructor that isn't in line with a couple of proposals for modifying the attributes of threads (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0484r1.pdf, http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0320r1.pdf; thanks @hkaiser).

On one hand adding the overload taking a thread pool won't stop us from adding one that takes attributes in the future, but on the other we could just add something like the attributes overload instead. It could allow setting a thread pool, priority, and stack size for now. @hkaiser I guess the important questions are how likely do you think it is that one of those papers will eventually be part of the standard in roughly that form, and what do we lose if we just add a constructor taking a thread pool explicitly?
